### DISCUSSION TITLE
Migrating a pre-existing dotfiles repository

### DISCUSSION THREAD

**Discussion by britter** [twpayne/chezmoi#2330] [Q&A] [answered]
Hi,

I've been managing [my dotfiles in a repository](https://github.com/britter/dotfiles) for years now, using a link script. So I have for example a `.bashrc` file in that repository, and then I clone the repository and link the file into my user home. I'm now considering migrating to chezmoi, but I want to keep the history of the repository and the history of the individual files. I wonder what the best process for doing that is and if there are any gotchas that I need to be aware of. Here's what I'm thinking of doing:

- Install chezmoi
- Clone my dotfiles repository to `~/.local/share/chezmoi`
- Rename files and dirs starting with a `.` to `dot_`
- Run `chezmoi diff` to see what would happen
- When satisfied run `chezmoi apply`
- Commit and push changes

Does this make sense? Anything else I need to initialize in the chezmoi repository that would normally be initialized by `chezmoi init`?

Thank you!

**Comment by garyo** [comment]
I came here just now to write exactly this! Interested to hear what the best practices are for migrating a long-existing dotfiles repo.

**Comment by twpayne** [comment, marked as answer]
> Does this make sense? 

Yes, your proposed series of steps sound very reasonable. git should detect the renames and so you'll maintain per-file history.

> Anything else I need to initialize in the chezmoi repository that would normally be initialized by chezmoi init?

No, nothing else is needed. `chezmoi init` just creates an empty git repo.

Another approach that you might want to consider if you want to do an incremental migration is to use [`.chezmoiroot`](https://www.chezmoi.io/reference/special-files-and-directories/chezmoiroot/) and [`chezmoi add --follow`](https://www.chezmoi.io/reference/commands/add/#-follow). `.chezmoiroot` allows you to keep the chezmoi-managed files separate from your link-script managed files (in your case this only matters for the `licenses` directory and `README.md` as chezmoi ignores everything beginning with a `.`). `chezmoi add --follow` tells chezmoi to add the targets of symlinks, rather than adding the symlink itself.

Together this means you'd do something like:

```console
$ git clone https://github.com/britter/dotfiles.git ~/.local/share/chezmoi
$ chezmoi cd
$ mkdir home
$ echo home > .chezmoiroot
$ chezmoi add --follow ~/.bashrc
$ rm .bashrc
```

...and similarly run `chezmoi add --follow ~/$FILE && rm $FILE` for the other dotfiles you want to chezmoi to manage.

This has the following advantages:
* The migration can be incremental.
* `chezmoi add` will pick up any extra attributes of the dotfile (e.g. if it's executable).
* At the end, you'll end up with your dotfiles in the `home` subdirectory in your dotfiles repo, so the root of your dotfiles repo will be uncluttered.

Hope this helps - please do follow up with any questions.

**Reply by britter** [reply]
I will give this a try. Thank you so much!

**Reply by bvdwalt** [reply]
This even helped me today. Thank you for the steps.

**Comment by garyo** [comment]
Thanks! I ended up just doing `chezmoi init` and then adding my existing repo as an upstream and pulling it all in. (Just in case chezmoi was going to populate the repo with anything like a .gitignore or anything else). That worked fine. But since `chezmoi init` creates an empty repo, your method above would be simpler.
